Index: share/prim_advance_mod.F90
===================================================================
--- share/prim_advance_mod.F90	(revision 1211)
+++ share/prim_advance_mod.F90	(working copy)
@@ -1635,7 +1635,7 @@
   use derivative_mod, only : derivative_t, divergence_sphere, gradient_sphere, vorticity_sphere
   use edge_mod, only : edgevpack, edgevunpack
   use bndry_mod, only : bndry_exchangev
-  use control_mod, only : moisture, qsplit, subcycle
+  use control_mod, only : moisture, qsplit, compute_mean_flux
   use hybvcoord_mod, only : hvcoord_t
 
   use physical_constants, only : cp, cpwater_vapor, Rgas, kappa
@@ -1739,7 +1739,7 @@
         ! ================================
         ! Accumulate mean Vel_rho flux in vn0
         ! ================================
-        if(subcycle)then
+        if(compute_mean_flux)then
         do j=1,nv
            do i=1,nv
               elem(ie)%derived%vn0(i,j,:,k)=elem(ie)%derived%vn0(i,j,:,k)+eta_ave_w*vtemp(i,j,:)
